include/curl/easy.h

```diff
@@ -25,6 +25,98 @@
 extern "C" {
 #endif
 
+/* Application can use a memory PKCS12 certificate with CURLOPT_SSLCERT
+ *  with OpenSSL, Schannel or SecTransport
+ *
+ *
+ *  Example if with certificate binary data and size in:
+ *    void* certdata;
+ *    size_t certsize;
+ *    ...
+ *  set the certificate with:
+ *    struct curl_blob structblob;
+ *    curl_init_blob_dup(&structblob, certdata, certsize);
+ *    curl_easy_setopt(curl, CURLOPT_KEYPASSWD, &structblob);
+ *             my_setopt_str(curl, CURLOPT_SSLCERT, &structblob);
+ *  struct curl_blob is just a 0x20 bytes structures which will
+ *  start bt a magic string and contain data
+ *
+ *  with curl_init_blob_dup, certdata can be discarded after
+ *    calling curl_easy_setopt
+ *  with curl_init_blob_persist, certdata need be valid when the
+ *  connexion is made
+ *
+ * Note : never call curl_init_blob_xxx without a pointer to a
+ *   valid 0x20 bytes block
+ * And never use the string CURL_BLOB_MAGIC without a correct
+ *   blob block
+ *
+ *
+ * about the idea, see https://curl.haxx.se/mail/lib-2016-09/0074.html */
+
+#define CURL_BLOB_MAGIC          ("\x01" "CurlMemBlob" "\xff\x01")
+#define CURL_BLOB_MAGIC_STRLEN   (14)
+#define CURL_BLOB_MAGIC_SIZE     (CURL_BLOB_MAGIC_STRLEN + 1)
+#define CURL_BLOB_OFFSET_DUPFLAG (CURL_BLOB_MAGIC_SIZE)
+#define CURL_BLOB_SIZE_DUPFLAG   (1)
+#define CURL_BLOB_OFFSET_DATALEN (CURL_BLOB_OFFSET_DUPFLAG + \
+                                  CURL_BLOB_SIZE_DUPFLAG)
+#define CURL_BLOB_SIZE_DATALEN   (sizeof(size_t))
+#define CURL_BLOB_OFFSET_DATAPTR (CURL_BLOB_OFFSET_DATALEN + \
+                                  CURL_BLOB_SIZE_DATALEN)
+#define CURL_BLOB_SIZE_DATAPTR   (sizeof(void *))
+#define CURL_BLOB_SIZE           (CURL_BLOB_OFFSET_DATAPTR + \
+                                  CURL_BLOB_SIZE_DATAPTR)
+
+#define CURL_BLOB_DUPFLAG_COPY   (1)
+#define CURL_BLOB_DUPFLAG_NOCOPY (0)
+
+/* the struct curl_blob store binary data parameters
+ * the structure size is 0x20 and can be followed by binary data */
+struct curl_blob {
+  char blob_internal[CURL_BLOB_SIZE];
```

Comment:
With the new `_BLOB` approach, is there a reason anymore why this isn't just a regular struct with individual fields for the values ? Would also remove a lot of the need for the magic things here...

Reply:
In the different vtls code (openssl, windows schannel, apple sectransport), we only have one variable data->set.ssl.cert . Then we call a funciton which determine if this pointer is a string filename or a blob.

I wrote several version of the blob system, which all worked, trying to be compatible with what you want.
I understand that : you want coherecy in the library . But now I'm not sure of what you want.

I think the better way is probablmy you modify the code yourself to be as you think it is better